Series.value_counts: Preserve original ordering

tomspur opened the pull request:
Ensure that value_counts returns the same ordering of the indices than the input object
when sorting the values no matter if it is ascending or descending.

- [x] closes #12679
- [x] tests added / passed
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry
- [ ] About 10 tests of the testsuite are currently failing and I hope to find a fix for them soon as well. Wanted to put up this PR now already to get feedback on the current implementation as well.

jreback commented on pandas/core/series.py:
```diff
@@ -2660,16 +2661,6 @@ def sort_values(self, axis=0, ascending=True, inplace=False,
             raise ValueError("This Series is a view of some other array, to "
                              "sort in-place you must create a copy")
 
-        def _try_kind_sort(arr):
```
why you changing this?

tomspur replied:
I couldn't get it working to separate the sorting and ascending vs non-ascending ordering further below in the code and changed it to pass it on to `nargsort` that does both at once

jreback commented on pandas/core/algorithms.py:
```diff
@@ -706,6 +706,17 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
+            # Use same index as the original values
+            if result.index.isna().sum() > 0:
+                fill_value = result[result.index.isna()].values[0]
+                result = result.reindex(unique(values), fill_value=fill_value)
```
huh? this is way beyond scope

tomspur replied:
The problem is that `_value_counts_arraylike` replaces all generic nan values (e.g. `None`) with `np.NaN` and reindexing has a mismatch between the original `None` and the new `np.NaN` in the index. The fill_value replaces the new missing value with the original value again.

Alternatively, it is also possible to use a generic `values = values.fillna(np.NaN)` to replace the `None`s with `np.NaN`s if there are nans in the index. I'll push a follow up commit about this in a bit. Would that be better than the above?

tomspur commented on pandas/tests/series/test_value_counts.py:
```diff
@@ -0,0 +1,58 @@
+    tm.assert_series_equal(Series(vc.index), s)
+
+
+def test_original_ordering_value_counts2():
```
Thanks! I'll move them to that location as well.

TomAugspurger commented on pandas/core/algorithms.py:
```diff
@@ -706,6 +706,19 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
             result = Series(counts, index=keys, name=name)
 
+            # Use same index as the original values
+            if result.index.isna().sum() > 0:
```
I don't understand this section. Are there cases where `values.fillna(np.nan)` does anything?

tomspur replied:
What basically happens a few lines further is a reindex of a value_counts:
```
vals = [1, None, 2, 2]
vals.value_counts(dropna=False).reindex(pd.Series(pd.unique([1, None, 2, 2])))
```
The value_counts transforms the None into NaN in the index, so that the count will be NaN as well (and this value count is removed from the returning reindexed Series).

The fillna is used to transform the None from the reindex as well, so that this is executed instead, so that the resulting Series still contains the value count:
```
pd.Series(vals).value_counts(dropna=False).reindex(pd.Series(pd.unique(vals)).fillna(np.NaN)
```

I hope this makes sense, because I am quite confused as well from times. If you have another solution to fix this edge case, please let me know...

jreback replied:
@tomspur I am not sure what you are doing here at all. see the OP for the solution, which is a simple re-index. this PR is way out of scope.

tomspur replied:
@jreback 
The re-index seems to work for `sort=False` and non-categorical indices and I could add that in this PR.

I wanted to further fix this for `sort=True`, when several values have the same count, e.g. this from the tests that were added in this PR:
```
In [2]: Series(list('bacaef'))
In [3]: s.value_counts(sort=True)                                                                                                                                                                                                             
Out[3]: 
a    2
c    1
e    1
b    1
f    1
dtype: int64
```
And I wanted to have them in the order `abcef` as well in that case.

I'll have a look then at `sort=False` for now as you suggested and leave `sort=True` for another PR

jreback reviewed (changes requested):
see comments

tomspur committed:
Series.value_counts: Preserve original ordering when using sort=False

Ensure that value_counts returns the same ordering of the indices than the input object
when sorting the values no matter if it is ascending or descending.

This fixes #12679.

tomspur committed:
Only reindex value_counts if bins is None

tomspur committed:
Modularize value_counts tests and xfail sort=True ones

tomspur commented on pandas/tests/test_algos.py:
```diff
@@ -962,6 +962,42 @@ def test_value_counts_uint64(self):
+        # vc = s.value_counts(sort=True, ascending=False)
+        # tm.assert_series_equal(Series(vc.index), ref_sorted)
+        # vc = s.value_counts(sort=True, ascending=True)
+        # tm.assert_series_equal(Series(vc.index), ref_sorted)
```
I implemented this as suggested and the `sort=True` cases are `xfail`ing now as well

tomspur commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1634,6 +1634,7 @@ Other
 
 - Bug where C variables were declared with external linkage causing import errors if certain other C libraries were imported before Pandas. (:issue:`24113`)
 - Require at least 0.28.2 version of ``cython`` to support read-only memoryviews (:issue:`21688`)
+- :meth:`Series.value_counts` returns the counts in the same ordering as the original series when using ``sort=False``
```
Added the issue number here as well

jreback commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -1634,6 +1634,7 @@ Other
 
 - Bug where C variables were declared with external linkage causing import errors if certain other C libraries were imported before Pandas. (:issue:`24113`)
 - Require at least 0.28.2 version of ``cython`` to support read-only memoryviews (:issue:`21688`)
+- :meth:`Series.value_counts` returns the counts in the same ordering as the original series when using ``sort=False`` (:issue:`12679`)
```
move to api breaking changes

tomspur replied:
Done and pushed

jreback commented on pandas/core/algorithms.py:
```diff
@@ -708,6 +708,10 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
         result = result.sort_values(ascending=ascending)
+    elif bins is None:
+        uniq = unique(values)
+        if not isinstance(result.index, CategoricalIndex):
```
why is this check needed? (or maybe it just needs to be for an *ordered* categorical)

tomspur replied:
The issue was the test `TestCategoricalSeriesAnalytics.test_value_counts`:
```
cats = Categorical(list('abcccb'), categories=list('cabd'))
s = Series(cats, name='xxx')
res = s.value_counts(sort=False)
```
which returns `0` for the `d`category as well, which is not in the `unique(values)`. Is there another possibility to get access to that initial categorical from the index?

jreback commented on pandas/core/algorithms.py:
```diff
@@ -708,6 +708,10 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
         result = result.sort_values(ascending=ascending)
+    elif bins is None:
+        uniq = unique(values)
+        if not isinstance(result.index, CategoricalIndex):
```
you can use ABCCategoricalIndex

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -962,6 +962,52 @@ def test_value_counts_uint64(self):
         if not compat.is_platform_32bit():
             tm.assert_series_equal(result, expected)
 
+    def test_value_counts_nonsorted_single_occurance(self):
```
paramterize on sort

tomspur replied:
Done and pushed

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -962,6 +962,52 @@ def test_value_counts_uint64(self):
+        vc = s.value_counts(sort=False, ascending=True)
+        tm.assert_series_equal(Series(vc.index), s)
+
+    @pytest.mark.xfail(reason="sort=True does not guarantee the same order")
```
why is this xfail?

tomspur replied:
I could not get it working for the `sort=True` case and left the tests only for possible future fixing... Would you prefer deleting them and adding them when `sort=True` works as well?

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -962,6 +962,52 @@ def test_value_counts_uint64(self):
+        vc = s.value_counts(sort=True, ascending=True)
+        tm.assert_series_equal(Series(vc.index), s)
+
+    def test_value_counts_nonsorted_double_occurance(self):
```
parametrize these.

tomspur replied:
The `double_occurance` tests have different expected results and I wouldn't parametrize it due to that. Or would you do this as well in that case?

jreback replied:
I c, ok then, parameterize over ascending though

tomspur committed:
Move whatsnew entry to correct place

tomspur committed:
Use ABCCategoricalIndex instead of CategoricalIndex

jreback commented on pandas/core/algorithms.py:
```diff
@@ -708,6 +708,10 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
     if sort:
         result = result.sort_values(ascending=ascending)
+    elif bins is None:
+        uniq = unique(values)
```
so do
``uniq = uniques(values)`` for the non-EA case (above)
and do ``uniq = Series(values)._values.unique()`` for the EA case. though this means computing it twice. maybe have to work on that.

tomspur replied:
I added it above, although it is not computed twice now. Did you mean it like this?

tomspur committed:
Calculate unique for the EA vs non-EA case

tomspur committed:
Parametrize value_counts test

jreback commented on pandas/core/algorithms.py:
```diff
@@ -698,16 +699,21 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
     if sort:
         result = result.sort_values(ascending=ascending)
+    elif bins is None:
+        if not isinstance(result.index, ABCCategoricalIndex):
```
this should not be necessary

tomspur replied:
It unfortunately is for cases, where the categories contain more elements than the values: https://github.com/pandas-dev/pandas/pull/24302#discussion_r243770038

jreback replied:
it is not necessary otherwise the impl is incorrect

categoricals uniq already handles this

tomspur replied:
It doesn't seem so as `unique` does not contain `d` in this example:
```

In [31]: s = pd.Series(pd.Categorical(list('baabc'), categories=list('abcd')))                                                                                                                                                                

In [32]: s.value_counts()                                                                                                                                                                                                                     
Out[32]: 
b    2
a    2
c    1
d    0
dtype: int64

In [33]: pd.unique(s)                                                                                                                                                                                                                         
Out[33]: 
[b, a, c]
Categories (3, object): [b, a, c]
```

tomspur replied:
https://github.com/pandas-dev/pandas/blob/master/pandas/core/arrays/categorical.py#L2267 also mentions that unused categories are not returned.

It would be consistent to not return it in the `value_counts` in the above example as well, but would change the current behaviour... What do you think?

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -962,6 +962,46 @@ def test_value_counts_uint64(self):
+    @pytest.mark.parametrize("sort", [True, False])
+    def test_value_counts_single_occurance(self, sort):
+        # GH 12679
+        if sort == True:
```
make this a param with mark=pytest.mark.xfail

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -962,6 +962,46 @@ def test_value_counts_uint64(self):
+        # the same order.
+        s = Series(list('bacdef'))
+
+        # Guarantee the same index if value_counts(sort=False) is used
```
parametrize over ascending as well

use 
```
result=
expected=
tm.assert_series_equal(result, expected)
```

jreback commented on pandas/tests/test_algos.py:
```diff
@@ -962,6 +962,46 @@ def test_value_counts_uint64(self):
+        vc = s.value_counts(sort=False, ascending=True)
+        tm.assert_series_equal(Series(vc.index), ref_nonsorted)
+
+    @pytest.mark.xfail(reason="sort=True does not guarantee the same order")
```
see comment above

jreback commented on pandas/core/algorithms.py:
```diff
@@ -698,16 +699,21 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
             result = Series(values)._values.value_counts(dropna=dropna)
```
wrap ``Series(values)`` only once

WillAyd commented on doc/source/whatsnew/v0.24.0.rst:
```diff
@@ -385,6 +385,7 @@ Backwards incompatible API changes
 - :func:`read_csv` will now raise a ``ValueError`` if a column with missing values is declared as having dtype ``bool`` (:issue:`20591`)
 - The column order of the resultant :class:`DataFrame` from :meth:`MultiIndex.to_frame` is now guaranteed to match the :attr:`MultiIndex.names` order. (:issue:`22420`)
 - :func:`pd.offsets.generate_range` argument ``time_rule`` has been removed; use ``offset`` instead (:issue:`24157`)
+- :meth:`Series.value_counts` returns the counts in the same ordering as the original series when using ``sort=False`` (:issue:`12679`)
```
Move to 0.25

WillAyd reviewed (changes requested):
Can you merge master?

WillAyd commented:
Closing as stale. Ping if you'd like to continue
